ddtrace/contrib/cherrypy/middleware.py

```diff
@@ -0,0 +1,139 @@
+        # the priority. See: https://docs.cherrypy.org/en/latest/extend.html#tools-ordering
+        cherrypy.Tool.__init__(self, "on_start_resource", self._on_start_resource, priority=95)
+
+    def _setup(self):
+        cherrypy.Tool._setup(self)
+        cherrypy.request.hooks.attach("on_end_request", self._on_end_request, priority=5)
+        cherrypy.request.hooks.attach("after_error_response", self._after_error_response, priority=5)
+
+    def _on_start_resource(self):
+        if config.cherrypy.get("distributed_tracing", False) or self._use_distributed_tracing:
+            propagator = HTTPPropagator()
+            context = propagator.extract(cherrypy.request.headers)
+            # Only need to active the new context if something was propagated
+            if context.trace_id:
+                self._tracer.context_provider.activate(context)
+
+        cherrypy.request._datadog_span = self._tracer.trace(
+            SPAN_NAME,
+            service=trace_utils.int_service(None, config.cherrypy, default=self._service),
+            span_type=SpanTypes.WEB,
+        )
+
+    def _after_error_response(self):
+        span = getattr(cherrypy.request, "_datadog_span", None)
+
+        if not span:
+            log.warning("cherrypy: tracing tool after_error_response hook called, but no active span found")
+            return
+
+        if not span.sampled:
+            return
+
+        span.error = 1
+        span.set_tag(errors.ERROR_TYPE, cherrypy._cperror._exc_info()[0])
+        span.set_tag(errors.ERROR_MSG, str(cherrypy._cperror._exc_info()[1]))
+        span.set_tag(errors.STACK, cherrypy._cperror.format_exc())
+
+        self._close_span(span)
+
+    def _on_end_request(self):
+        span = getattr(cherrypy.request, "_datadog_span", None)
+
+        if not span:
+            log.warning("cherrypy: tracing tool on_end_request hook called, but no active span found")
+            return
+
+        if not span.sampled:
+            return
+
+        self._close_span(span)
+
+    def _close_span(self, span):
+        # Let users specify their own resource in middleware if they so desire.
+        # See case https://github.com/DataDog/dd-trace-py/issues/353
+        if span.resource == SPAN_NAME:
+            # In the future, mask virtual path components in a
+            # URL e.g. /dispatch/abc123 becomes /dispatch/{{test_value}}/
+            # Following investigation, this should be possible using
+            # [find_handler](https://docs.cherrypy.org/en/latest/_modules/cherrypy/_cpdispatch.html#Dispatcher.find_handler)
+            # but this may not be as easy as `cherrypy.request.dispatch.find_handler(cherrypy.request.path_info)` as
+            # this function only ever seems to return an empty list for the virtual path components.
+
+            # For now, default resource is method and path:
+            #   GET /
+            #   POST /save
```

Comment:
:+1: thanks for the write up! I think I had originally misread your comment in the other thread 😆. Yeah this is good enough for now. Part of me wonders if we could monkey-patch `find_handler` to do a `tracer.current_span()` and set the resource that way but I'm ok with leaving as-is for now to keep things simple.

Reply:
Yes, that could well work! After this, if I don't immediately start looking into it (curiosity kills cats/time), then I'll pull together some notes into a ticket so we have the details and commentary on the new feature in one place.